tests/test_raft.rs

```diff
@@ -529,53 +535,179 @@ fn test_progress_paused() {
+            (state, term)
+        };
+        if raft.state != exp_state {
+            panic!("#{}: state = {:?}, want {:?}", i, raft.state, exp_state);
         }
-        if raft.term != 1 {
-            panic!("#{}: term = {}, want {}", i, raft.term, 1)
+        if raft.term != exp_term {
+            panic!("#{}: term = {}, want {}", i, raft.term, exp_term)
         }
     }
 }
 
 #[test]
+fn test_leader_cycle_not_pre_vote() {
+    test_leader_cycle(false)
+}
+
+#[test]
+fn test_leader_cycle_pre_vote() {
+    test_leader_cycle(true)
+}
+
+// test_leader_cycle verifies that each node in a cluster can campaign
+// and be elected in turn. This ensures that elections (including
+// pre_vote) work when not starting from a clean state (as they do in
+// test_leader_election)
+fn test_leader_cycle(pre_vote: bool) {
+    let mut network = Network::new(vec![None, None, None], pre_vote);
+    for campaigner_id in 1..4 {
+        network.send(vec![new_message(campaigner_id, campaigner_id, MessageType::MsgHup, 0)]);
+        for sm in network.peers.values() {
+            if sm.id == campaigner_id && sm.state != StateRole::Leader {
+                panic!("preVote={}: campaigning node {} state = {:?}, want StateLeader",
+                       pre_vote,
+                       sm.id,
+                       sm.state);
+            } else if sm.id != campaigner_id && sm.state != StateRole::Follower {
+                panic!("preVote={}: after campaign of node {}, node {} had state = {:?}, want \
+                        StateFollower",
+                       pre_vote,
+                       campaigner_id,
+                       sm.id,
+                       sm.state);
+            }
+        }
+    }
+}
+
+#[test]
+fn test_vote_from_any_state() {
+    test_vote_from_any_state_(MessageType::MsgRequestVote);
+}
+
+#[test]
+fn test_prevote_from_any_state() {
+    test_vote_from_any_state_(MessageType::MsgRequestPreVote);
+}
+
+fn test_vote_from_any_state_(vt: MessageType) {
+    fn test_vote_from_any_state_impl(vt: MessageType, state: StateRole) {
+        let mut sm = new_test_raft(1, vec![1, 2, 3], 10, 1, new_storage());
+        sm.term = 1;
+        match state {
+            StateRole::Follower => {
+                let term = sm.term;
+                sm.become_follower(term, 3);
+            }
+            StateRole::PreCandidate => sm.become_precandidate(),
+            StateRole::Candidate => sm.become_candidate(),
+            StateRole::Leader => {
+                sm.become_candidate();
+                sm.become_leader();
+            }
+        }
+        // Note that setting our state above may have advanced sm.term
+        // past its initial value.
+        let orig_term = sm.term;
+        let new_term = sm.term + 1;
+        let mut msg = new_message(2, 1, vt, 1);
+        msg.set_term(new_term);
+        msg.set_log_term(new_term);
+        msg.set_index(42);
+        sm.step(msg).expect("");
+        assert_eq!(sm.msgs.len(), 1);
```

Comment:
The error logs should be kept here so that the programmer could easily target the error case, since this function is run many times with different states input. See the original PR.